mentat/embeddings.py

```diff
@@ -1,77 +1,107 @@
-import json
+import logging
 import os
-import sqlite3
-from pathlib import Path
 from timeit import default_timer
 
-import numpy as np
+import chromadb
+from chromadb.utils.embedding_functions import OpenAIEmbeddingFunction
 
 from mentat.code_feature import CodeFeature, count_feature_tokens
-from mentat.errors import MentatError
-from mentat.llm_api_handler import (
-    count_tokens,
-    model_context_size,
-    model_price_per_1000_tokens,
-)
+from mentat.llm_api_handler import model_price_per_1000_tokens
 from mentat.session_context import SESSION_CONTEXT
 from mentat.session_input import ask_yes_no
-from mentat.utils import mentat_dir_path, sha256
+from mentat.utils import mentat_dir_path
 
-EMBEDDINGS_API_BATCH_SIZE = 2048
+client = chromadb.PersistentClient(path=str(mentat_dir_path / "chroma"))
 
 
-class EmbeddingsDatabase:
-    def __init__(self, output_dir: Path | None = None):
-        self.output_dir = output_dir or mentat_dir_path
-        os.makedirs(self.output_dir, exist_ok=True)
-        self.path = Path(self.output_dir) / "embeddings.sqlite3"
-        self._connect()
+class Collection:
+    _collection = None
 
-    def _connect(self):
-        self.conn = sqlite3.connect(self.path)
-        with self.conn as db:
-            db.execute(
-                "CREATE TABLE IF NOT EXISTS embeddings "
-                "(checksum TEXT PRIMARY KEY, vector BLOB)"
-            )
-
-    def set(self, items: dict[str, list[float]]):
-        with self.conn as db:
-            db.executemany(
-                "INSERT OR REPLACE INTO embeddings (checksum, vector) VALUES (?, ?)",
-                [
-                    (key, sqlite3.Binary(json.dumps(value).encode("utf-8")))
-                    for key, value in items.items()
-                ],
-            )
-
-    def get(self, keys: list[str]) -> dict[str, list[float]]:
-        with self.conn as db:
-            cursor = db.execute(
-                "SELECT checksum, vector FROM embeddings WHERE checksum IN"
-                f" ({','.join(['?']*len(keys))})",
-                keys,
-            )
-            return {row[0]: json.loads(row[1]) for row in cursor.fetchall()}
-
-    def exists(self, key: str) -> bool:
-        with self.conn as db:
-            cursor = db.execute("SELECT 1 FROM embeddings WHERE checksum=?", (key,))
-            return cursor.fetchone() is not None
-
-    def __del__(self):
-        self.conn.close()
-
-
-database = EmbeddingsDatabase()
+    def __init__(self, embedding_model: str):
+        api_key = os.getenv("OPENAI_API_KEY")
+        # src: https://cookbook.openai.com/examples/vector_databases/chroma/using_chroma_for_embeddings_search
+        embedding_function = OpenAIEmbeddingFunction(
+            api_key=api_key, model_name=embedding_model
+        )
+        self._collection = client.get_or_create_collection(
+            name=f"mentat-{embedding_model}",
+            embedding_function=embedding_function,  # type: ignore
```

Comment:
The typing issue here and below is related to the OpenAIEmbeddingFunction, which is their in-house recommended approach but isn't setup to handle images (which the rest of them are).